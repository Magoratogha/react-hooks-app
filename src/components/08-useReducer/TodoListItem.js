import PropTypes from 'prop-types';
import { memo } from 'react';

const TodoListItem = memo(({ todo, index, handleToggle, handleDelete }) => {
    return (
        <li
            className="list-group-item">
            <p
                onClick={() => handleToggle(todo.id)}
                className={'mb-0 ' + (todo.isDone ? 'completed' : '')}>
                {index + 1}. {todo.description}
            </p>
            <button
                onClick={() => handleDelete(todo.id)}
                className="btn btn-outline-danger btn-sm">
                Delete
            </button>
        </li>
    );
});

TodoListItem.propTypes = {
    todo: PropTypes.object.isRequired,
    index: PropTypes.number.isRequired,
    handleToggle: PropTypes.func.isRequired,
    handleDelete: PropTypes.func.isRequired
};

export default TodoListItem;