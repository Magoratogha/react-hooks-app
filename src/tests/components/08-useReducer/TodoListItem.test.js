import { render, screen } from "@testing-library/react";
import TodoListItem from "../../../components/08-useReducer/TodoListItem";
import { todosMock } from "../../mocks/todos";
import userEvent from "@testing-library/user-event";

describe('Test on <TodoListItem />', () => {

    const user = userEvent.setup();
    const handleToggleMock = jest.fn();
    const handleDeleteMock = jest.fn();

    test('should render properly', () => {
        const { container } = render(<TodoListItem
            todo={todosMock[0]}
            index={2}
            handleToggle={handleToggleMock}
            handleDelete={handleDeleteMock} />);
        expect(container).toMatchSnapshot();
    });

    test('should call the delete function', async () => {
        render(<TodoListItem
            todo={todosMock[0]}
            index={2}
            handleToggle={handleToggleMock}
            handleDelete={handleDeleteMock} />);
        const deleteButton = screen.queryByText('Delete', { selector: 'button' });
        await user.click(deleteButton);
        expect(handleDeleteMock).toHaveBeenCalledWith(todosMock[0].id);
    });

    test('should call the toggle function', async () => {
        render(<TodoListItem
            todo={todosMock[0]}
            index={2}
            handleToggle={handleToggleMock}
            handleDelete={handleDeleteMock} />);
        const toggleParagraph = screen.queryByText('3. ' + todosMock[0].description, { selector: 'p' });
        await user.click(toggleParagraph);
        expect(handleToggleMock).toHaveBeenCalledWith(todosMock[0].id);
    });

    test('should render the text properly', () => {
        render(<TodoListItem
            todo={todosMock[0]}
            index={2}
            handleToggle={handleToggleMock}
            handleDelete={handleDeleteMock} />);
        expect(screen.getByText('3. ' + todosMock[0].description, { selector: 'p' })).toBeTruthy();
    });

    test('should have the complete class if isDone property is true', () => {
        const todoMock = { ...todosMock[0], isDone: true };
        render(<TodoListItem
            todo={todoMock}
            index={2}
            handleToggle={handleToggleMock}
            handleDelete={handleDeleteMock} />);
        expect(screen.getByText('3. ' + todosMock[0].description, { selector: 'p' })).toHaveClass('completed');
    });
});