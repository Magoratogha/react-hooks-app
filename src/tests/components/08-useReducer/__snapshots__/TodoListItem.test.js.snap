// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <TodoListItem /> should render properly 1`] = `
<div>
  <li
    class="list-group-item"
  >
    <p
      class="mb-0 "
    >
      3
      . 
      To learn React
    </p>
    <button
      class="btn btn-outline-danger btn-sm"
    >
      Delete
    </button>
  </li>
</div>
`;
